memory/memory/memory.py

```diff
@@ -0,0 +1,120 @@
+from hailtop.hail_logging import AccessLogger
+from hailtop.tls import get_in_cluster_server_ssl_context
+from hailtop.utils import AsyncWorkerPool, retry_transient_errors
+from gear import setup_aiohttp_session, rest_authenticated_users_only
+
+uvloop.install()
+
+BATCH_PODS_NAMESPACE = os.environ['HAIL_BATCH_PODS_NAMESPACE']
+log = logging.getLogger('batch')
+routes = web.RouteTableDef()
+
+socket = '/redis/redis.sock'
+
+@routes.get('/healthcheck')
+async def healthcheck(request):  # pylint: disable=unused-argument
+    return web.Response()
+
+
+@routes.get('/api/v1alpha/objects')
+@rest_authenticated_users_only
+async def get_object(request, userdata):
+    filename = request.query.get('q')
+    etag = request.query.get('etag')
+    userinfo = await get_or_add_user(request.app, userdata)
+    username = userdata['username']
+    log.info(f'memory: request for object {filename} from user {username}')
+    result = await get_file_or_none(request.app, username, userinfo, filename, etag)
+    if result is None:
+        raise web.HTTPNotFound()
+    etag, body = result
+    return web.Response(headers={'ETag': etag}, body=body)
+
+
+async def get_or_add_user(app, userdata):
+    users = app['users']
+    username = userdata['username']
+    if username not in users:
+        k8s_client = app['k8s_client']
+        gsa_key_secret = await retry_transient_errors(
+            k8s_client.read_namespaced_secret,
+            userdata['gsa_key_secret_name'],
+            BATCH_PODS_NAMESPACE,
+            _request_timeout=5.0)
+        gsa_key = base64.b64decode(gsa_key_secret.data['key.json']).decode()
+        users[username] = {'fs': GCS(blocking_pool=app['thread_pool'], key=json.loads(gsa_key))}
+    return users[username]
+
+
+def make_redis_key(username, filepath):
+    return f'{ username }_{ filepath }'
+
+
+async def get_file_or_none(app, username, userinfo, filepath, etag):
+    filekey = make_redis_key(username, filepath)
+    fs = userinfo['fs']
+
+    cached_etag = await app['redis_pool'].execute('HGET', filekey, 'etag', encoding='ascii')
+    if cached_etag is not None and cached_etag == etag:
+        result = await app['redis_pool'].execute('HGET', filekey, 'body')
+        log.info(f"memory: Retrieved file {filepath} for user {username} with etag'{etag}'")
+        return cached_etag, result
+
+    log.info(f"memory: Couldn't retrieve file {filepath} for user {username}: current version not in cache.")
+    if filekey not in app['files_in_progress']:
+        try:
+            log.info(f"memory: Loading {filepath} to cache for user {username}")
+            app['worker_pool'].call_nowait(load_file, app['redis_pool'], app['files_in_progress'], filekey, fs, filepath)
+        except asyncio.QueueFull:
+            pass
+    return None
+
+
+async def load_file(redis, files, file_key, fs, filepath):
+    try:
+        files.add(file_key)
+        data = await fs.read_binary_gs_file(filepath)
+        etag = await fs.get_etag(filepath)
```

Comment:
This is also a race condition but as you pointed out on my FS PR, this will be fixed by making the etag available on the read object stream.

Reply:
Yep! This is exactly why I wanted the etag to be available and I'll fix this when I port to the new python aiostorage stuff.